core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -136,19 +151,25 @@ private boolean checkTasks() {
         }
       }
 
-      return !closed && (tasks.hasNext() || hasRunningTask);
+      return !closed.get() && (tasks.hasNext() || hasRunningTask);
     }
 
-    private Future<?> submitNextTask() {
-      if (!closed && tasks.hasNext()) {
-        return workerPool.submit(tasks.next());
+    private Future<Optional<Task<T>>> submitNextTask() {
+      if (!closed.get()) {
+        if (!yieldedTasks.isEmpty()) {
+          return workerPool.submit(yieldedTasks.removeFirst());
+        }
```

Comment:
nit: combine this into 
If / else if

to avoid the implicit If else if caused by the dual returns

Reply:
done!